### PR TITLE
Feature/only unlock if process dead

### PR DESCRIPTION
**Pull request by JerryShea**

### CODE REVIEW

**Review by RobAustin — CHANGES_REQUESTED** [review]
we discussed this in the dev meeting, and the consensus was that we should consider other approaches than throwing an UnrecoverableTimeoutException() and forcing the user to run some other code ( such as Unlock Main ) to remove the lock.

**Comment by JerryShea** [comment]
Let's discuss at dev meeting again this week. Only behaviour change introduced in this PR is if the new `queue.lock.recover.dead.process` parameter is set, and then we behave consistently with the append lock i.e. we unlock if the locking process is dead, otherwise throw exception

**Review by RobAustin — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/queue/impl/single/TableStoreWriteLock.java [hunk 1]
```diff
@@ -88,11 +88,16 @@ public void lock() {
...
-            tlPauser.reset();
-            lock();
+            else if (onlyUnlockIfProcessDead) {
+                if (forceUnlockIfProcessIsDead())
+                    lock();
+                else
+                    throw new UnrecoverableTimeoutException(new IllegalStateException(warningMsg + UNLOCK_MAIN_MSG));
+            } else {
+                warn().on(getClass(), warningMsg + UNLOCKING_FORCIBLY_MSG);
+                forceUnlock(value);
```
Comment (nicktindall) [inline src/main/java/net/openhft/chronicle/queue/impl/single/TableStoreWriteLock.java]: I feel like the value variable could be named better (e.g. currentLockValue or currentPID or whatever it is)

Comment (nicktindall) [reply src/main/java/net/openhft/chronicle/queue/impl/single/TableStoreWriteLock.java]: Especially now we're increasing the number of lines between where it's assigned and where it's used

**Review by nicktindall — APPROVED** [review]
I like this change. Perhaps it should become default in .23?

Although if we do that maybe we should have config like `forceUnlockOnTimeoutWhen` = `[ALWAYS | NEVER | LOCKING_PROCESS_DEAD]` rather than having `queue.lock.recover.dead.process` and `queue.dont.recover.lock.timeout` as separate and possibly inconsistent confgs

**Comment by JerryShea** [comment]
> I like this change. Perhaps it should become default in .23?
> 
> Although if we do that maybe we should have config like `forceUnlockOnTimeoutWhen` = `[ALWAYS | NEVER | LOCKING_PROCESS_DEAD]` rather than having `queue.lock.recover.dead.process` and `queue.dont.recover.lock.timeout` as separate and possibly inconsistent confgs

We normally give users two releases before we change things like this, but happy to take votes from the group.

I have implemented the enum which was very helpful - it helped shake out some inconsistencies (and further reinforces the case to refactor these locking classes)

File: src/main/java/net/openhft/chronicle/queue/impl/single/TSQueueLock.java [hunk 2]
```diff
@@ -122,15 +121,25 @@ public void waitForLock() {
         }
     }
 
+    private void warnAndForceUnlock(String msg, long value) {
+        warnLock(msg, value);
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/queue/impl/single/TSQueueLock.java:125]: duplicated but not quite the same code - we should address with https://github.com/OpenHFT/Chronicle-Queue/issues/994

File: src/test/java/net/openhft/chronicle/queue/impl/single/TSQueueLockTest.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+import java.util.Collection;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.*;
+
+public class TSQueueLockTest extends QueueTestCommon {
```
Comment (JerryShea) [inline src/test/java/net/openhft/chronicle/queue/impl/single/TSQueueLockTest.java:32]: this is copy/pasted from TableStoreWriteLockTest. This test definitely needs some more work/refactoring - see https://github.com/OpenHFT/Chronicle-Queue/issues/994

**Pull request merged**